### DISCUSSION TITLE
Const Groupings Versus Creating User Inconviences

### DISCUSSION THREAD

**Discussion by Wajinn** [vlang/v#22282] [General]
Why would V go down this path of creating more user inconveniences on this issue? This includes forcing users to rewrite and update their code unnecessarily or seemingly frivolously. Please reconsider this issue and keep in mind V's many loyal fans and attracting new users. I've happen to have already read some comments concerning such (groupings and needing to do rewrites).

Groupings just makes more sense than unnecessarily typing the same keyword repeatedly. Especially if V is promoting readability and ease of use. V is such a great language, that it should stay looking nice.

This excessiveness (which is arguably a kind of violation of DRY)...
```
const a = 1
const b = 2
const c = 3
const d = 4
const e = 'more'
```
Would arguably look better as...
```
const (
a = 1
b = 2
c = 3
d = 4
e = 'more'
)
```

If the argument is to force only 1 way, then this can be done by typing `const (a = 1)` for single items, where `()` is always used. In fact, that's arguably the direction that import and __global should go as well.

Understand there can be reasons and are totally sympathetic to developers, but if Go and other languages have figured out how to do this, why would V go backwards? And finally, medvednikov typed he [preferred const groupings](https://github.com/vlang/v/discussions/16568#discussioncomment-4285778), so not understanding as to what could have happened.

**Comment by medvednikov** [comment]
There's no re-writing, vfmt updates const and import groups to the new syntax.

It's much easier to grep for consts and imports this way.

**Reply by Wajinn** [reply]
First of all, team V has been doing wonderfully, so it's rare that I'm in disagreement about any particular direction or choice. But on this point, feel compelled to bring it up.

> There's no re-writing, vfmt updates const and import groups to the new syntax.
> It's much easier to grep for consts and imports this way.

There are several issues with the present update and changes, that perhaps is not being fully appreciated, from the perspective of users or casuals. 

**1)** While `import` may be only a few lines,` const` can be a very large number of lines.

The point of why languages have used `()` for large numbers of consts, appears to have been forgotten or not appreciated. Many people are not so keen to type or look at "const" for 20 or more lines nor are hardcore or pro developers. 

**2)** While some may view banning groups "()" as a kind of advantage, aesthetically, it is completely not so.

Competing languages of V, have no issues with giving users groups. Where V appears to be going backwards, by taking a user facing convenience away. The aesthetic component of this, can greatly affect many new and casual users.

**3)** While it is being claimed that it's easier for some to grep consts and imports, other languages do not have this issue.

The problem with the "easier to grep" statement, has and is being disputed. Furthermore,  other competing languages don't have this problem, so why does V? For example, import and const groups don't appear to be an issue in Go, so then how and why did it become such a major problem for V? Is it really a problem or is it actually a preference by some? 

**4)** V gives the warning, "const () groups will be an error after 2025-01-01 (`v fmt -w source.v` will fix that for you)"

The point I'm hoping to bring light to, is the accumulation of changes over time. It is of course a balance of any language needing changes and evolving, with how those changes affect users and general adoption

The const `()` groups was not an issue previously in V, but forcibly became so. Making the way V was written before, now outdated, and many programs and libraries written in V have not and various will not do the update (some because not 1.0 yet and others moving on). This can be confirmed by many outdated libraries (though new ones being created too), and I'm such a person to check.

If it has not been noticed, many older libraries of V do not get updated. Arguably because of constant changes, even if some of them can be fixed by v fmt. And the older the library becomes, the more changes are needed, than simply running v fmt.

**5)** The do it "only 1 way rule", is arguably better satisfied by always using `()`

This has been pointed out by not just me, but other supporters, such as JoanaBLate. As in typing `const (a = 1)`, would satisfy the rule, in a way similar to `${}` being the only way for string interpolation. And in arguing for only `${}`, it was chosen because it was the more flexible and powerful version. Along those lines of logic, `const ()` can be used for one item or multiple.

**Reply by medvednikov** [reply]
You make good points. It's always hard with decisions like this because half of the users prefers one thing, the other half prefers the opposite. And you have to find the balance.

Grepping for imports in Go is probably easier because imports are strings, so `rg '"io"'` will most likely give you what you're looking for.

`rg io` will give you millions of actual `io` calls. So I think having only `import foo` is correct.

I've launched a twitter poll:

https://x.com/v_language/status/1838202005992464881
https://x.com/v_language/status/1838201264900907048

Not the most scientific method, but it can give a good picture.

**Reply by JalonSolov** [reply]
And that is why it is an issue.  Suppose V goes back to using () for const values.  If you're grep'ing for something named `foo`, there is no way to know if it's a const or not without opening every file where `foo` was found, and looking to see if it is inside the `const ()`.

Whereas if every const line starts with `const`, you'll immediately see that in the grep output.

**Reply by medvednikov** [reply]
True.

On the other hand there's also added verbosity from "pub const", not just "const":

![image](https://github.com/user-attachments/assets/06a7b053-a229-4bbc-b701-93b99ebd4f1d)

**Reply by JalonSolov** [reply]
Most are fine with a small amount of verbosity, if it enhances clarity.

**Reply by Wajinn** [reply]
> If you're grep'ing for something named foo...

The problem with the argument of "easier to grep for consts", is that it is possible to write regular expressions in grep to find words within parentheses or other English grammatical marks. [Example 1](https://unix.stackexchange.com/questions/103004/grep-regex-only-for-match-anything-between-parenthesis), [Example 2](https://unix.stackexchange.com/questions/33697/finding-text-between-two-specific-characters-or-strings)... 

It then, on the surface, gives the appearance of being a specific preference versus it being truly necessary to globally remove an existing feature for all users.

The level of visual clutter, verbosity, and excessive typing is debatably exponential, in comparison to a simple and single `()`, that already existed in V (and Go). Along those lines, it comes across as hurting V's goal of being simple and its aesthetic appeal, when compared to competition.

**Reply by JalonSolov** [reply]
Yes, it is possible to grep for something inside `()` in the code... but if you don't know that you need to do that, how does it help?  If you just grep for `foo` because you have no idea whether it is a const or a global or a regular variable, does that mean you have to do 3 separate greps now to avoid opening every file where it appears, just to find out which type it might be?

**Reply by Wajinn** [reply]
Many users would use various kinds of convenient applications to grep for whatever, versus feeling confined to only the cmd line or terminal. Alternativeto.net, shows a large number of [examples](https://alternativeto.net/software/grep/). Beyond that, it should be possible to create a specific grep solution, to return the kind of results desired.

The grep problem, appears too individually specific or resolvable in various other ways, versus global removal of an existing feature in the language.

**Reply by medvednikov** [reply]
I agree, that we can have something like `v grep` which will use the upcoming pure V language server, and that's it, problem solved.

**Reply by medvednikov** [reply]
Or to make it even more explicit `v find const foo`, `v find import bar` etc

**Reply by JalonSolov** [reply]
As long as it's easy to find what I want to find, the syntax doesn't matter to me.

**Reply by Wajinn** [reply]
`v grep` or `v find` looks to be a very nice and useful solution, that would prevent splitting fans of the language on this subject. It is likely best to go in that direction, as it can be seen as "adding value" to using V, that can attract more users. As oppose to removing a perceived value or existing feature, that can possibly upset or divide fans of language.

**Reply by Wajinn** [reply]
> I've launched a twitter poll:
> https://x.com/v_language/status/1838202005992464881 https://x.com/v_language/status/1838201264900907048
> Not the most scientific method, but it can give a good picture.

Think that polls are a good idea, as a way to gauge public sentiment about topics. However, echoing a comment made on twitter, various people voting on the poll may not be actual V users or fans of the language. The more people who are not users are voting, the more it might mangle perception or distort the results.

Luckily, I do recall it being mentioned somewhere, that team V (developers and supporters) has a significant share or are as large of a factor in the decision.

**Reply by medvednikov** [reply]
By the way we already have `v where`, so just need to add that functionality.

As expected, very different opinions:

<img width="652" alt="image" src="https://github.com/user-attachments/assets/c53aa1f6-beb0-4a5c-aaa1-c87ee97b13ef">

<img width="663" alt="image" src="https://github.com/user-attachments/assets/d46974e7-c0a9-4697-b085-f46751d42aac">

It's like spaces vs tabs. Everyone has their preference.

**Reply by jorgeluismireles** [reply]
> By the way we already have `v where`, so just need to add that functionality.

Is nice to have this tool! Thanks!

In different modules I have several `enum ComputerSet` and several `pub const vid = ...`.

- This works ok, giving all the related enums by the filter:
```
$ v version
V 0.4.7 4ee948f
$ v where enum Computer* -dir .
...
...
```
- This fails (gives a panic)
```
$ v where enum *Set -dir .
V panic: 
query: .*enum\s**Set\s*.*
err  : --------^
ERROR: err_syntax_error
...
```
- This gives nothing, I think it should gives whole list:
```
v where enum * -dir .
```
- This gives nothing (seems const is not programmed yet, const vid really exists even is pub):
```
v where const vid -dir .
```
Is odd, the last fails if I have:
```
const vid := 0x1a86
```
But it works ok if I have:
```
const vid = 0x1a86
```

**Reply by JalonSolov** [reply]
The current `v where` just uses regex patterns.  The change Alex is talking about will use the LSP to do a full and proper search based on the parsed files.

For example, `const vid := 0x1a86` isn't valid V.  `v fmt` would change it for you, to the correct version of `const vid = 0x1a86` as you say works.  The new thing would also parse it correctly, as it would do the same substitution as `v fmt`.  A simple regex search, though, can't do that.

And don't bother saying "but you could make the regex work" - there's no reason to even try, with the improvements that are coming.

**Reply by jorgeluismireles** [reply]
Exactly `const vid := 0x1a86` works currently and also `const vid = 0x1a86`. It think it should force in the compiler the former is rejected, so there was a "single way to write constants".

**Reply by JalonSolov** [reply]
I believe the plan is to force only the 2nd way, but not yet - don't want to break people using `:=`.

**Reply by JalonSolov** [reply]
Although there have also been arguments for making `const vid :=` the "correct" way, because then it matches regular var initialization, such as `x := 0`.

**Comment by spytheman** [comment]
Groupings impede simple search tools like grep and ripgrep.

**Reply by Wajinn** [reply]
Per medvednikov, thought that members of team V had a viable solution of using `v grep`, `v find`, or `v where` (please see earlier posts). This looked to be a compromise solution, where neither side had to lose or be frustrated.

The hope is that V can reduce or avoid going down the road of enshittification, as has happened to various other popular languages, where high levels of unnecessary user facing verbosity was created. Even if it is perceived that such continual levels of additional verbosity are justified, the perception between users and developers can be different.

**Reply by spytheman** [reply]
I am a user of the language too, and that dichotomy that you implied is false.

`v where` sucks, compared to grep and rg (in speed and uniformity of the UX). I suggest using it, and comparing for yourself.

**Reply by Wajinn** [reply]
First, you have my utmost respect, so there is no intention to be otherwise or cause harm.

As for, `v where`, disagree that it sucks. But of course, coming from a user's perspective. `v where` is doing what the help says it would. When looking for something **within** a const grouping, it finds it (see below).

I'm not understanding why not improve the  `v where` tool (as a compromise and for the future of V) versus take away functionality that many users and supporters prefer. It kind of looks like specific persons of prominence have some unspecified requirement, so rather strip V of an already existing functionality.

In regards to grep and ripgrep, they are just search tools like `v where`, which looks like it can be modified to solve this unspecified requirement, and allow a compromise where everybody wins. The tool ripgrep is open-source MIT, would be shocked if whatever particular thing it does that's needed, could not be translated to V. Go also has a few grep and grep-like knockoffs.

```

v where const time_sleep -dir /examples -v

Find: const time_sleep | visibility: all mutability: any
within directories: ['/examples'] 1 matches Found

/examples/demo_2.v
13 : [ time_sleep = 500 ]
```

**Reply by spytheman** [reply]
Tools like grep and ripgrep, do not need to understand/parse the content they are processing, and that gives them a huge speed advantage. Tools like `v where` do need to parse the content. As a minimum they have to keep state of when a group has been started already, and when it is finished, in order to know if the current line is part of it. That takes time and complexity -> they can not be as fast as non context sensitive tools.

This limits them in answering questions like 'quickly find all constants in a module', or 'find all public symbols on a module' etc. Without grouping, those are very easy and fast to answer, even with non specialized tools like grep and ripgrep.

**Reply by spytheman** [reply]
> Go also has a few grep and grep-like knockoffs.

They also suck, for the same reason - the grouping syntax just makes certain workflows a lot harder, for not much good reason, beside the aesthetic preferences of some.

**Reply by Wajinn** [reply]
Speed differences within seconds or fractions of a second, are arguably irrelevant to the task of general users finding what they are looking for, which `v where` can do (find what's in a const grouping). 

If a user has some other specific requirements or preferences, that `v where` can't satisfy or be modified to address, they could still use those other external tools. In other words, `v where` and ripgrep (or other tools of a different ecosystem), satisfies both groups versus the creation of division. Cutting off supporters, with either abrupt or arbitrary user facing changes, appears contrary to V's claims or goals of stability, gives a bad look, and narrows the audience it can appeal to. 

In regards to ripgrep or other, if a person is already satisfied with using some other tool to find something, it doesn't appear to make sense to then turn around and strip out existing conveniences or preferred functionality in the language, used by a large number (around 45%) of its users. Unless there is something more going on or unelaborated critical necessity.

The attraction of V, for the wider audience, would likely be its built-in conveniences that other languages do not have, V created tools (including the fun of making them), often mentioned readability, etc... That should arguably include being less verbose. To continually take away things in which users and supporters like about V, could decrease its popularity, in an already crowded and very hostile environment. And that seems counter-productive.

**Reply by Wajinn** [reply]
> They also suck, for the same reason - the grouping syntax just makes certain workflows a lot harder, for not much good reason, beside the aesthetic preferences of some.

Yet Go kept grouping syntax in the language. That naturally generates the question of why they did so. They could have just as easily and forcibly, pulled it out. It could be that they made a cost–benefit analysis, where keeping the grouping syntax for users had **won**.

**Reply by jorgeluismireles** [reply]
IMHO Golang is different in the sense they use the pattern for "almost everything", next is valid golang code.
```go
import (
	"fmt"
)

var (
	a int
	b float32
)

const (
	c = 10
	d = 32.43
)

type (
	e struct{}
	f interface{}
)
```
At least, V cannot take advantage of the pattern too much because has no global vars nor the structs are defined using `type` keyword.

**Reply by Wajinn** [reply]
V had used this pattern more, but it was stripped away. `import` used to have it. It was a compromise of the community (or it had appeared to be) to just remove it from `import` and keep it for `const` and `__global` (used with flag), but now it looks to be under siege too.

**Comment by spytheman** [comment]
> Go and other languages have figured out how to do this, why would V go backwards?

They have not. It is a pain to search for stuff in Go projects (for me, I do not know about others), using language agnostic tools.
